tensortrade/environments/trading_environment.py

```diff
@@ -102,6 +102,7 @@ def __init__(self,
         self._observation_lows = kwargs.get('observation_lows', -np.iinfo(np.int32).max)
         self._observation_highs = kwargs.get('observation_highs', np.iinfo(np.int32).max)
         self._max_allowed_loss = kwargs.get('max_allowed_loss', 0.1)
+        self._drop_n_frames = kwargs.get('drop_n_frames', 0)
```

Comment:
`self._drop_n_frames` and `self._rewind` should be a part of `data_feed`, not `trading_environment`.

Reply:
Then in `feed.reset()`, you can call `self._rewind()`

Reply:
Yes I agree, the changes belong in the data feed class. As I was making the `DataFeed` I thought about this issue, but I only saw it being a problem for live trading, which would not be done for a bit. Some kind of fast forward method can be put into the `DataFeed` class and then on creation it can fast forward itself by calling `next`. In the case of live trading, it would then have to wait until  being completely fast forwarded. After that creation of the environment could move on as is.

Reply:
that's how I implemented it for the first time. Let's assume I have the following `build_env` function:
```
   def build_env(self, df):
      exchange = Exchange("exchange", service=execute_order)(
         self.exchange_stream(df)
      )
      feed = DataFeed([self.feed_module(df)])

      portfolio = Portfolio(USD, [
         Wallet(exchange, 10000 * USD),
         Wallet(exchange, 10 * BTC),
      ])
      env = TradingEnvironment(
         feed=feed,
         portfolio=portfolio,
         action_scheme='managed-risk',
         reward_scheme='risk-adjusted',
         price_history=self.price_history(df),
         window_size=4,
         drop_n_frames=self._drop_n_frames,
         renderers = [self._chart_renderer, self._file_logger]
      )
```

The data frame is used in three different places: 
* It goes to Exchange:    

      def exchange_stream(self, df):
          return Stream("USD-BTC", df[self._price_column].tolist())

* It goes to DataFeed: 

   ```
   def feed_module(self, df):
      with Module("data") as coinbase_ns:
         nodes = {name: Stream(name, df[name].tolist()) for name in df.columns}

      return coinbase_ns
    ```

* It goes to price_history for the renderer

       def price_history(self, df):
          return df[self._drop_n_frames:].reset_index().rename(columns={'index':'datetime'})

So you can see, that DataFeed is only one of three potential places. If DataFrame implements `rewind()` than other parts can be left untouched which will cause desynchronization.

The ideal solution is to make all three places outputs from a pipeline with a single input. So we feed data either from OHLCV or a live exchange, select a price column and pass to `Exchange`, convert `datetime` index and pass to `price_history` and apply all transformations, drop some original columns and pass to `DataFeed`.

This way if `DataFeed` rewinds data, it will be rewound in all three places. @mwbrulhardt will this work? 

Another way is to create a `Rewind(Node)` node so people can put it in the middle of the pipeline. But I don't know how to rewind correctly so all ascendants and descendants remain in-sync. There are so many "similar" methods: `next()`, `forward()`, `gather()` etc.

That's why I made a quicker solution, and put it into TradingEnvironment, since it has `step` method already and knows how to use all three different data sources together. But I'd like to implement it right, I need more help about DataFeed implementations